clippy_lints/src/lifetimes.rs

```diff
@@ -71,38 +71,34 @@ enum RefLt {
     Named(Name),
 }
 
-fn bound_lifetimes(bound: &TyParamBound) -> Option<HirVec<&Lifetime>> {
-    if let TraitTyParamBound(ref trait_ref, _) = *bound {
-        let lt = trait_ref.trait_ref
-                          .path
-                          .segments
-                          .last()
-                          .expect("a path must have at least one segment")
-                          .parameters
-                          .lifetimes();
-
-        Some(lt)
-    } else {
-        None
-    }
-}
-
 fn check_fn_inner(cx: &LateContext, decl: &FnDecl, generics: &Generics, span: Span) {
-    if in_external_macro(cx, span) || has_where_lifetimes(cx, &generics.where_clause) {
+    if in_external_macro(cx, span) || where_has_lifetime(cx, &generics.where_clause) {
         return;
     }
 
     let bounds_lts = generics.ty_params
                              .iter()
-                             .flat_map(|ref typ| typ.bounds.iter().filter_map(bound_lifetimes).flat_map(|lts| lts));
+                             .flat_map(|ref typ| typ.bounds.iter().flat_map(|bound| {
+                                 if let TraitTyParamBound(ref trait_ref, _) = *bound {
+                                     trait_ref.trait_ref
+                                              .path
+                                              .segments
+                                              .last()
+                                              .expect("a path must have at least one segment")
+                                              .parameters
+                                              .lifetimes()
+                                 } else {
+                                     HirVec::new()
+                                 }
+                             }));
```

Comment:
Why removing the function here? I like closures for small stuffs but not so much for longer functions.